packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx

```diff
@@ -0,0 +1,222 @@
+import { Jsonifiable } from 'type-fest';
+
+export type LengthFunction = ((text: string) => number) | ((text: string) => Promise<number>);
+
+export interface RedactedFuncionResponseMetadata {
+  isRedacted: true;
+  chunks: string[];
+}
+
+const getEncoder = _.once(() => getEncoding('cl100k_base'));
+
+function openAITokenCount(text: string) {
+  return getEncoder().encode(text).length;
+}
+
+export async function TruncateByChars(
+  {
+    children,
+    maxLength,
+  }: {
+    children: AI.Node;
+    maxLength: number;
+  },
+  { render }: AI.ComponentContext
+) {
+  const stringified = await render(children);
+  if (stringified.length <= maxLength) {
+    return stringified;
+  }
+  return `${stringified.slice(0, maxLength - 3)}...`;
+}
+
+export interface LargeFunctionResponseProps {
+  maxLength: number;
+  failedMaxLength: number;
+  numChunks: number;
+  lengthFunction?: LengthFunction;
+}
+
+async function LargeFunctionResponseHandler(
+  {
+    children,
+    maxLength = 4000,
+    failedMaxLength = 1000,
+    numChunks = 4,
+    // use Cohere token counter?
+    lengthFunction = openAITokenCount,
+    ...props
+  }: AI.PropsOfComponent<typeof FunctionResponse> & LargeFunctionResponseProps,
+  { render, logger, getContext }: AI.ComponentContext
+) {
+  if (props.failed) {
+    return (
+      // TODO: fix issue between maxLength chars and tokens
+      <FunctionResponse {...props}>
+        <TruncateByChars maxLength={failedMaxLength}>{children}</TruncateByChars>
+      </FunctionResponse>
+    );
+  }
+
+  let stringified = await render(children);
+
+  // Option 1: do nothing if it's already small enough
+  if ((await lengthFunction(stringified)) <= maxLength) {
+    return <FunctionResponse {...props}>{stringified}</FunctionResponse>;
+  }
+
+  stringified = yamlOptimizeIfPossible(stringified);
+
+  // Option 2: try dumping as YAML. If it's small enough, then we are done.
+  if ((await lengthFunction(stringified)) <= maxLength) {
+    return <FunctionResponse {...props}>{stringified}</FunctionResponse>;
+  }
+
+  // Option 3 (last reosrt): split into chunks and allow LLM to query by similarity
+  // Requires Cohere API key for doing similarity search
+  const cohereConfig = getContext(cohereContext);
+
+  if (!cohereConfig.api_key) {
+    // Not yet an error, but this is an error just waiting to happen
+    logger.warn(
+      { CohereContext: cohereConfig },
+      'FunctionResponse is too big, but Cohere API key is not set. Please set it in the context.'
+    );
+  }
+
+  const splitter = new RecursiveCharacterTextSplitter({
+    chunkSize: maxLength / numChunks,
+    chunkOverlap: maxLength / numChunks / 10,
+    lengthFunction,
+  });
+  const chunks = await splitter.splitText(stringified);
+
+  return (
+    <FunctionResponse {...props} metadata={{ isRedacted: true, chunks }}>
+      ... The response is too big and hence redacted. The response can be queried using semantic similarity search by
```

Comment:
Will having multiple such `<FunctionResponse>s` in the conversation be a problem?

Reply:
I don't think so. We only present the model with the last one, and LLM gets at least one turn to react to it (also the `loadBySimilarity` itself will not trigger redaction so it can be called repeatedly). Is there any other scenario you were thinking about?

Reply:
WDYM we only present the model with the last one -- won't this be persisted in the conversation? And re:side-by-side:

- user
- functionCall
- functionResponse
- functionCall
- functionResponse
- assistant

is certainly a valid scenario, but I'm likely misunderstanding something.

A reasonable answer might be just that the model is smart enough to infer that the last one takes precedence over previous ones, but I was curious if this is something that either can't happen (i.e. I'm misunderstanding things) or you've played around with and is fine.

Reply:
Ooh sorry yeah you're right. I'm assuming that the model is smart enough. I have certainly used multiple of these in the same conversation and didn't have any issues.

But this is a valid point. I'll play with the descriptions a bit more to make it a bit more obvious.

Reply:
I improved the function description. For now I don't have better ideas to make sure this doesn't happen. I say we can assume that it works until we see any issues.
I put a TODO in the code to track this concern.

WDYT?

Reply:
Works for me